pkg/controller/statefulset/stateful_set_control.go

```diff
@@ -607,6 +619,80 @@ func (ssc *defaultStatefulSetControl) updateStatefulSet(
 	return &status, nil
 }
 
+func updateStatefulSetAfterInvariantEstablished(
+	ctx context.Context,
+	ssc *defaultStatefulSetControl,
+	set *apps.StatefulSet,
+	replicas []*v1.Pod,
+	updateRevision *apps.ControllerRevision,
+	status apps.StatefulSetStatus,
+) (*apps.StatefulSetStatus, error) {
+
+	replicaCount := int(*set.Spec.Replicas)
+
+	// we compute the minimum ordinal of the target sequence for a destructive update based on the strategy.
+	updateMin := 0
+	maxUnavailable := 1
+	if set.Spec.UpdateStrategy.RollingUpdate != nil {
+		updateMin = int(*set.Spec.UpdateStrategy.RollingUpdate.Partition)
+
+		// if the feature was enabled and then later disabled, MaxUnavailable may have a value
+		// other than 1. Ignore the passed in value and Use maxUnavailable as 1 to enforce
+		// expected behavior when feature gate is not enabled.
+		var err error
+		maxUnavailable, err = intstrutil.GetValueFromIntOrPercent(intstrutil.ValueOrDefault(set.Spec.UpdateStrategy.RollingUpdate.MaxUnavailable, intstrutil.FromInt(1)), int(replicaCount), false)
+		if err != nil {
+			return &status, err
+		}
+
+	}
+
+	// Collect all targets in the range between the partition and Spec.Replicas. Count any targets in that range that are unhealthy
+	// i.e. terminated or not running and ready as unavailable). Select the (MaxUnavailable - Unavailable) Pods, in order with respect
+	// to their ordinal for termination. Delete those pods and count the successful deletions. Update the status with the correct
+	// number of deletions.
+	unavailablePods := 0
+	for target := len(replicas) - 1; target >= updateMin; target-- {
```

Comment:
If updateMin is not zero, this could miss counting some unhealthy pods?

Reply:
the invariant has been established till this point(https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/statefulset/stateful_set_control.go#L497-L501), that all pods are running and ready before we attempt any deletion. See https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/statefulset/stateful_set_control.go#L444 and  https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/statefulset/stateful_set_control.go#L452
and https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/statefulset/stateful_set_control.go#L497-L501 so technically we could leave this check as `target >=updateMin` since we from nowards are only deleting pods from replicas to updateMin(partition) in the next loop. For extra security, i will make it 0. @lavalamp